### PR TITLE
[agent] Add Support to Spark Connect

### PR DESCRIPTION
**Pull request by skcoirz**
Spark Connect allows computation delegation to clusters though DataFrame API. It has different data type. Trying to re-use as much as possible here. (spark connect: https://spark.apache.org/docs/latest/spark-connect-overview.html)
- example notebook added

### CODE REVIEW

File: langchain/agents/agent_toolkits/spark/base.py [hunk 1]
```diff
@@ -36,6 +36,82 @@ def create_spark_dataframe_agent(
...
+) -> AgentExecutor:
+    """Construct a spark agent from an LLM and dataframe."""
+    try:
+        from pyspark.sql.connect.dataframe import DataFrame
+    except ImportError:
+        raise ValueError(
+            "spark package not found, please install with `pip install pyspark`"
+        )
+
+    if not isinstance(df, DataFrame):
```
Comment (amaliujia) [inline langchain/agents/agent_toolkits/spark/base.py:79]: This check may not be needed? Spark Connect DataFrame should be stable enough now it is in that package and named as `DataFrame`.

Comment (skcoirz) [reply langchain/agents/agent_toolkits/spark/base.py:79]: I couldn’t find `DataFrame` in `pyspark.sql.connect`, it’s only visible in `..connect.dataframe`. Is it what you mentioned above? (the latter part)

I think the typecheck is a good thing to have as the input type is `Any`...

Comment (amaliujia) [reply langchain/agents/agent_toolkits/spark/base.py:79]: I was referring this https://github.com/apache/spark/blob/1b54b014543801c98d86e6e1151c7b763f98cfde/python/pyspark/sql/connect/dataframe.py#L94

**Review by hwchase17 — COMMENTED** [review]
imo probably better to just have one notebook for both of them? are they really different enough to deserve separate notebooks?

**Comment by skcoirz** [comment]
> imo probably better to just have one notebook for both of them? are they really different enough to deserve separate notebooks?

@hwchase17 , for langchain, they are similar except the data type, but for the spark users, they are handled differently. One is local, the other is remotely calling clusters. For the Spark Connect, it also supports its own interpreter. I'm thinking of making langchain available inside of their cluster interpreter (need to figure out an elegant way to handle dependencies)

File: langchain/agents/agent_toolkits/spark/base.py [hunk 2]
```diff
@@ -36,6 +36,82 @@ def create_spark_dataframe_agent(
...
+        return_intermediate_steps,
+        max_iterations,
+        max_execution_time,
+        early_stopping_method,
+        agent_executor_kwargs,
+        **(kwargs or {}),
+    )
+
+
+def create_spark_connect_dataframe_agent(
```
Comment (allisonwang-db) [inline langchain/agents/agent_toolkits/spark/base.py:56]: So the only difference between this function and `create_spark_dataframe_agent` is the validation for the input dataframe type. I think we can merge these two functions. Users should use the same method to create a Spark agent, regardless of whether spark connect is enabled.

**Review by HyukjinKwon — APPROVED** [review]

**Review by gengliangwang — APPROVED** [review]

**Review by amaliujia — COMMENTED** [review]
LGTM

**Review by amaliujia — COMMENTED** [review]
LGTM

**Pull request merged**